shared/src/commonMain/kotlin/org/hyperskill/app/step/presentation/StepActionDispatcher.kt

```diff
@@ -26,16 +45,70 @@ class StepActionDispatcher(
                     .fold(
                         onSuccess = {
                             sentryInteractor.finishTransaction(sentryTransaction)
-                            onNewMessage(Message.StepLoaded.Success(it))
+                            onNewMessage(
+                                Message.StepLoaded.Success(
+                                    step = it,
+                                    practiceStatus = if (action.stepRoute is StepRoute.Learn) {
+                                        PracticeStatus.AVAILABLE
+                                    } else {
+                                        PracticeStatus.UNAVAILABLE
+                                    }
+                                )
+                            )
                         },
                         onFailure = {
                             sentryInteractor.finishTransaction(sentryTransaction, throwable = it)
                             onNewMessage(Message.StepLoaded.Error)
                         }
                     )
             }
+            is Action.FetchNextStepQuiz -> {
+                if (action.currentStep.type == Step.Type.THEORY && !action.currentStep.isCompleted) {
+                    stepInteractor.completeStep(action.currentStep.id)
+                }
+
+                var nextRecommendedStep = stepInteractor
+                    .getNextRecommendedStepByTopicId(action.currentStep.topic)
+                    .getOrElse {
+                        onNextStepQuizFetchedError(action.currentStep)
+                        return
+                    }
+
+                while (!BlockName.supportedBlocksNames.contains(nextRecommendedStep.block.name) && nextRecommendedStep.canSkip) {
+                    stepInteractor
+                        .skipStep(nextRecommendedStep.id)
+                        .getOrElse {
+                            onNextStepQuizFetchedError(action.currentStep)
+                            return
+                        }
+
+                    nextRecommendedStep = stepInteractor
+                        .getNextRecommendedStepByTopicId(action.currentStep.topic)
+                        .getOrElse {
+                            onNextStepQuizFetchedError(action.currentStep)
+                            return
+                        }
+                }
```

Comment:
In my opinion, it could be a part of `StepInteractor`. `StepInteractor.getNextRecommendedStepByTopicId` can encapsulate this logic.